dtsync/publish.go

```diff
@@ -0,0 +1,161 @@
+	startHeadPublisher(host, topic, headPublisher)
+
+	return &publisher{
+		cancelPubSub:  cancel,
+		dtManager:     dtManager,
+		headPublisher: headPublisher,
+		host:          host,
+		tmpDir:        tmpDir,
+		topic:         t,
+	}, nil
+}
+
+func startHeadPublisher(host host.Host, topic string, headPublisher *head.Publisher) {
+	go func() {
+		log.Infof("Starting head publisher on peer ID %s for topic %s", host.ID(), topic)
+		err := headPublisher.Serve(host, topic)
+		if err != http.ErrServerClosed {
+			log.Errorf("Error head publisher stopped serving on peer ID %s for topic %s: %s", host.ID(), topic, err)
+		}
+		log.Infof("Stopped head publisher on peer ID %s for topic %s", host.ID(), topic)
+	}()
+}
+
+// NewPublisherFromExisting instantiates go-legs publishing on an existing
+// data transfer instance
+func NewPublisherFromExisting(dtManager dt.Manager, host host.Host, topic string, lsys ipld.LinkSystem) (*publisher, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	t, err := gpubsub.MakePubsub(ctx, host, topic)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	err = configureDataTransferForLegs(context.Background(), dtManager, lsys)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	headPublisher := head.NewPublisher()
+	startHeadPublisher(host, topic, headPublisher)
+
+	return &publisher{
+		cancelPubSub:  cancel,
+		headPublisher: headPublisher,
+		host:          host,
+		topic:         t,
+	}, nil
+}
+
+func (lp *publisher) UpdateRoot(ctx context.Context, c cid.Cid) error {
+	return lp.UpdateRootWithAddrs(ctx, c, lp.host.Addrs())
+}
+
+func (lp *publisher) UpdateRootWithAddrs(ctx context.Context, c cid.Cid, addrs []ma.Multiaddr) error {
+	if c == cid.Undef {
+		return errors.New("cannot update to an undefined cid")
+	}
+
+	log.Debugf("Publishing CID and addresses in pubsub channel: %s", c)
+	var errs error
+	err := lp.headPublisher.UpdateRoot(ctx, c)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	msg := Message{
+		Cid:   c,
+		Addrs: addrs,
+	}
+	err = lp.topic.Publish(ctx, EncodeMessage(msg))
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	return errs
+}
+
+func (lp *publisher) Close() error {
+	var errs error
+	lp.closeOnce.Do(func() {
+		err := lp.headPublisher.Close()
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+
+		// If tmpDir is non-empty, that means the publisher started the dtManager and
+		// it is ok to stop is and clean up the tmpDir.
+		if lp.tmpDir != "" {
+			ctx, cancel := context.WithTimeout(context.Background(), shutdownTime)
+			defer cancel()
+
+			err = lp.dtManager.Stop(ctx)
+			if err != nil {
+				log.Errorf("Failed to stop datatransfer manager: %s", err)
+				errs = multierror.Append(errs, err)
+			}
+			if err = os.RemoveAll(lp.tmpDir); err != nil {
+				log.Errorf("Failed to remove temp dir: %s", err)
+				errs = multierror.Append(errs, err)
+			}
+		}
+
+		t := time.AfterFunc(shutdownTime, lp.cancelPubSub)
```

Comment:
Why do we delay here?

Reply:
To give it time for `Close` to finish (shutdown gracefully) before canceling it.